src/librustc_metadata/index_builder.rs

```diff
@@ -245,44 +261,73 @@ impl<T> DepGraphRead for FromId<T> {
 pub struct EntryBuilder<'a, 'b: 'a, 'tcx: 'b> {
     pub tcx: TyCtxt<'b, 'tcx, 'tcx>,
     ecx: &'a mut EncodeContext<'b, 'tcx>,
-    hasher: StableHasher<Fingerprint>,
-    hcx: StableHashingContext<'b, 'tcx>,
+    hcx: Option<(StableHashingContext<'b, 'tcx>, StableHasher<Fingerprint>)>,
 }
 
 impl<'a, 'b: 'a, 'tcx: 'b> EntryBuilder<'a, 'b, 'tcx> {
 
     pub fn finish(self, def_id: DefId) {
-        let hash = self.hasher.finish();
-        self.ecx.metadata_hashes.push((def_id.index, hash));
+        if let Some((_, hasher)) = self.hcx {
+            let hash = hasher.finish();
+            self.ecx.metadata_hashes.push(EncodedMetadataHash {
+                def_index: def_id.index,
+                hash: hash,
+            });
+        }
     }
 
     pub fn lazy<T>(&mut self, value: &T) -> Lazy<T>
         where T: Encodable + HashStable<StableHashingContext<'b, 'tcx>>
     {
-        value.hash_stable(&mut self.hcx, &mut self.hasher);
+        if let Some((ref mut hcx, ref mut hasher)) = self.hcx {
+            value.hash_stable(hcx, hasher);
+            debug!("metadata-hash: {:?}", hasher);
+        }
         self.ecx.lazy(value)
     }
 
     pub fn lazy_seq<I, T>(&mut self, iter: I) -> LazySeq<T>
         where I: IntoIterator<Item = T>,
               T: Encodable + HashStable<StableHashingContext<'b, 'tcx>>
     {
-        let items: Vec<T> = iter.into_iter().collect();
-        items.hash_stable(&mut self.hcx, &mut self.hasher);
-        self.ecx.lazy_seq(items)
+        if let Some((ref mut hcx, ref mut hasher)) = self.hcx {
+            let iter = iter.into_iter();
+            let (lower_bound, upper_bound) = iter.size_hint();
+
+            if upper_bound == Some(lower_bound) {
+                lower_bound.hash_stable(hcx, hasher);
+                let ret = self.ecx.lazy_seq(iter.inspect(|item| item.hash_stable(hcx, hasher)));
+                debug!("metadata-hash: {:?}", hasher);
+                ret
```

Comment:
You should assert that the number of times the `inspect` callback gets called is equal to `lower_bound`.